### PR TITLE
Perf drawFramwork

### PR DESCRIPTION
**Pull request by etpinard**
Here are two perf commits in `Cartesian.drawFramework` that will help splom subplot generation. :racehorse: 

At 400 subplots (or 20 splom dimensions :smirk: ), these commits are able to speed up `drawFramework` from ~1000ms down under 200ms.

Note that it might be nice to make the new `joinLayer` from 7d881f6 a standard `Lib` function and replace all our `selectAll().data([0])` with it. Waiting on @alexcjohnson before going through with it.

### CODE REVIEW

**Comment by alexcjohnson** [comment]
Wow, I would not have guessed that those two were such a big drain, impressive find! Yes please, 🔪 all the `selectAll(...).data([0])`! The name `joinLayer` has never meant much to me though... I suppose I'd get used to it, but what about something like `Lib.ensureSingle`?

**Comment by etpinard** [comment]
> I suppose I'd get used to it, but what about something like Lib.ensureSingle

I like the sound of `ensureSingle`. Thanks!

File: src/components/colorbar/draw.js [hunk 1]
```diff
@@ -245,10 +245,9 @@ module.exports = function draw(gd, id) {
         cbAxisOut.setScale();
 
         // now draw the elements
-        var container = fullLayout._infolayer.selectAll('g.' + id).data([0]);
-        container.enter().append('g').classed(id, true)
-            .classed(cn.colorbar, true)
-            .each(function() {
+        var container = Lib.ensureSingle(fullLayout._infolayer, 'g', id, function(s) {
+            s.classed(cn.colorbar, true);
+            s.each(function() {
```
Comment (alexcjohnson) [inline src/components/colorbar/draw.js]: you don't like chaining here? ie
```js
s.classed(...)
    .each(...)
```

Comment (etpinard) [reply src/components/colorbar/draw.js]: Good catch. Thanks!

File: test/jasmine/tests/click_test.js [hunk 2]
```diff
@@ -418,8 +418,8 @@ describe('Test click interactions:', function() {
             var node = document.querySelector('rect.nwdrag');
             var pos = getRectCenter(node);
 
-            expect(node.classList[0]).toBe('drag');
-            expect(node.classList[1]).toBe('nwdrag');
+            expect(node.classList[1]).toBe('drag');
+            expect(node.classList[0]).toBe('nwdrag');
```
Comment (alexcjohnson) [inline test/jasmine/tests/click_test.js]: OK as is, but this seems ripe for a custom matcher like:
```
expect(node).toBeClassed(['drag', 'nwdrag', 'cursor-nw-resize']);
```
That would be order-independent but also verify these are the *only* classes applied.

File: src/plot_api/plot_api.js [hunk 3]
```diff
@@ -245,9 +245,7 @@ exports.plot = function(gd, data, layout, config) {
                 .attr('height', fullLayout.height);
         }
 
-        return Lib.syncOrAsync([
-            subroutines.layoutStyles
-        ], gd);
```
Comment (alexcjohnson) [inline src/plot_api/plot_api.js]: Too funny, so it just works! We may never know what obviated this...

File: src/plot_api/plot_api.js [hunk 4]
```diff
@@ -245,9 +245,7 @@ exports.plot = function(gd, data, layout, config) {
...
+        return Plots.previousPromises(gd);
```
Comment (etpinard) [inline src/plot_api/plot_api.js]: @alexcjohnson tests are :white_check_mark: after removing the extra `layoutStyles` call. Good riddance :boat:

File: src/plot_api/plot_api.js (hunk 3 above)
Comment (etpinard) [reply src/plot_api/plot_api.js]: _just works_<sup>TM</sup>

File: test/jasmine/tests/click_test.js (hunk 2 above)
Comment (etpinard) [reply test/jasmine/tests/click_test.js]: Good call.

`cartesian_test.js` uses assertions like:

```js
expect(fills[0].classList.contains('js-tozero')).toBe(true)
```

which could work here too, but yeah checking adding a custom assertion that test multiple class names at once should be worthwhile.

File: src/components/colorbar/draw.js (hunk 1 above)
Comment (etpinard) [reply src/components/colorbar/draw.js]: done in -> 8e6032d

File: test/jasmine/tests/click_test.js (hunk 2 above)
Comment (etpinard) [reply test/jasmine/tests/click_test.js]: done in b974eab

**Comment by alexcjohnson** [comment]
Beautiful! 💃 💃 💃

**Comment by etpinard** [comment]
Ok. This probably deserves to be in the next minor release `1.36.0`, so I'll wait a few more days in case we need to make another patch release in `1.35.x` series.

**Pull request merged**